File: scripts/export_openapi.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from dere_daemon.main import app


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


def _write_json(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(
        json.dumps(payload, indent=2, sort_keys=True, ensure_ascii=True) + "\n",
        encoding="utf-8",
    )


def main() -> None:
    schema = app.openapi()
    output_path = _repo_root() / "schemas" / "openapi" / "dere_daemon.openapi.json"
    _write_json(output_path, schema)
    print(f"Exported OpenAPI schema: {output_path.relative_to(_repo_root())}")


if __name__ == "__main__":
    main()

File: scripts/export_openapi.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

type JsonRecord = Record<string, unknown>;

const scriptDir = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(scriptDir, "..");

async function writeJson(path: string, payload: JsonRecord): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, `${JSON.stringify(payload, null, 2)}\n`, "utf-8");
}

async function main(): Promise<void> {
  const schemaPath = join(repoRoot, "schemas", "openapi", "dere_daemon.openapi.json");
  const raw = await readFile(schemaPath, "utf-8");
  const schema = JSON.parse(raw) as JsonRecord;
  await writeJson(schemaPath, schema);
  console.log(`Normalized OpenAPI schema: ${schemaPath.slice(repoRoot.length + 1)}`);
}

await main();
